Cannot use labeled argument assignment in html quotations

This code fails
```ocaml
let tcol_in_row ?(classes="col-xs-12 col-md-6") cols =
  let inrow x = <:html< <div class=$str:classes$>$x$</div> >> in
  <:html< <div class="row">$list:List.map inrow cols$</div> >>

let narrow_in_row cols =
  <:html< $tcol_in_row ~classes:"col-xs-1" cols$ >>
```
with the message
```
[ERROR] tcol_in_row ~classes is not a valid tag.
Allowed tags are [opt|int|flo|str|uri|list|alist|attrs] or the empty one.File "htmlutil.ml", line 9, characters 9-26:
While expanding quotation "html" in a position of "expr":
  Camlp4: Uncaught exception: Parsing.Parse_error
```

I'm sorry that I can't submit a patch; I am completely new to Camlp4.

BTW, I think that ocaml-cow (and the libraries it relies on) is great. I just finished using it to generate our new team website (http://parkas.di.ens.fr/). The generator (~1300 LoOC) slurps up a bunch of markdown and csv files and spits out html.

On 11 Feb 2016, at 08:02, Timothy Bourke <notifications@github.com> wrote:
> 
> [ERROR] tcol_in_row ~classes is not a valid tag

Yeah, unfortunately the ~ causes problems in the parser, so avoiding labelled arguments is a workaround for now.

--- a/CHANGES
+++ b/CHANGES
@@ -1,3 +1,7 @@
+1.4.1 (trunk):
+* Fix XML and HTML labeled argument assignment antiquotation syntax bug (#86)
+* Fix CSS space-less antiquotation syntax bug
+
 1.4.0 (27-Sep-2015):
 * Improve compatability with Type_conv >= 113.00 by renaming some of the
   syntax parser modules to be less generically named.

--- a/syntax/css/quotations.ml
+++ b/syntax/css/quotations.ml
@@ -20,14 +20,19 @@ module Q = Syntax.Quotation
 module AQ = Syntax.AntiquotSyntax
 
 let destruct_aq s =
-  try
-    let pos = String.index s ':' in
-    let space = String.index s ' ' in
-    if space < pos then raise Not_found;
+  (* If there is a space, it changes how we interpret antiquote classes. *)
+  let space_pos = try Some (String.index s ' ') with Not_found -> None in
+  let split pos =
     let len = String.length s in
     let name = String.sub s 0 pos
     and code = String.sub s (pos + 1) (len - pos - 1) in
     name, code
+  in
+  try
+    let pos = String.index s ':' in
+    match space_pos with
+    | Some spos -> if spos < pos then ("", s) else split pos
+    | None -> split pos
   with Not_found ->
     "", s
 

--- a/syntax/css/quotations.ml
+++ b/syntax/css/quotations.ml
@@ -52,7 +52,7 @@ object
             | "prop" -> <:expr< Css.Props $e$ >>
             | "" -> e
             | t ->
-              Printf.eprintf "[ERROR] \"%s\" is not a valid tag. Valid tags are [expr|prop]\n" t;
+              Printf.eprintf "[ERROR] \"%s\" is not a valid tag. Valid tags are [int|float|str|expr|prop]\n" t;
               Loc.raise _loc Parsing.Parse_error
           end
       | e -> super#expr e

--- a/syntax/xml/quotation.ml
+++ b/syntax/xml/quotation.ml
@@ -20,12 +20,19 @@ module Q = Syntax.Quotation
 module AQ = Syntax.AntiquotSyntax
 
 let destruct_aq s =
-  try
-    let pos = String.index s ':' in
+  (* If there is a space, it changes how we interpret antiquote classes. *)
+  let space_pos = try Some (String.index s ' ') with Not_found -> None in
+  let split pos =
     let len = String.length s in
     let name = String.sub s 0 pos
     and code = String.sub s (pos + 1) (len - pos - 1) in
     name, code
+  in
+  try
+    let pos = String.index s ':' in
+    match space_pos with
+    | Some spos -> if spos < pos then ("", s) else split pos
+    | None -> split pos
   with Not_found ->
     "", s
 

--- a/tests/render.ml
+++ b/tests/render.ml
@@ -42,6 +42,15 @@ let xml_expanders = [
   <:xml<$make_xml ~label:()$>>, make_xml_result;
 ]
 
+let css_int () = 5
+let css_int_result = string_of_int (css_int ())
+
+(* TODO: add real CSS tests not just these camlp4 antiquot parser tests *)
+let css_expanders = [
+  "css", <:css<&>>, "";
+  "css_fn", <:css<body { height: $int:css_int()$; } >>, css_int_result;
+]
+
 let suite name decl prefix =
   let suite = List.fold_left (fun acc (v, n, x, s) ->
       let aux t xml =

--- a/tests/render.ml
+++ b/tests/render.ml
@@ -21,6 +21,9 @@ let attrs = <:xml<hello="world" class="ab">>
 let attrs' = <:xml<hello="world"  class="ab">>
 let attrs'' = <:xml<hello="world" class="a b">>
 
+let make_xml ~label = <:xml<<tag/>&>>
+let make_xml_result = "<tag/>"
+
 let xml_expanders = [
   true, "opt",  <:xml<<z>$opt:opt$</z>&>>, "<z><a>a</a></z>";
   true, "opt'", <:xml<<z>$opt:opt'$</z>&>>, "<z/>";
@@ -33,6 +36,10 @@ let xml_expanders = [
   true, "attrs", <:xml<<tag $attrs:attrs$/>&>>, "<tag hello=\"world\" class=\"ab\"/>";
   true, "attrs'", <:xml<<tag $attrs:attrs'$/>&>>, "<tag hello=\"world\" class=\"ab\"/>";
   true, "attrs''", <:xml<<tag $attrs:attrs''$/>&>>, "<tag hello=\"world\" class=\"a b\"/>";
+  true, "labeled_xml_fn",
+  (let label = () in <:xml<$make_xml ~label$>>), make_xml_result;
+  true, "label_xml_fn_assignment",
+  <:xml<$make_xml ~label:()$>>, make_xml_result;
 ]
 
 let suite name decl prefix =

